### PR TITLE
Milestone #9: Add Authentication With Devise

### PR DESCRIPTION
**Pull request by carlylechia**
#### To complete this milestone, my partner and I did the following tasks:

- [x] Removed the `current_user` method in **ApplicationController**, devise will provide us with one.

- [x] [Installed and setup devise](https://github.com/heartcombo/devise#getting-started), and used it to make sure our app:

-      Can register a new user.

-      Users log in with a combination of email and password.

-      Users' hashed passwords are stored in the database.

-      Asks for confirmation of email.

-      Can reset password.

We also 

- [x] [Modified the views in devise](https://github.com/heartcombo/devise#configuring-views) for registration and login to match [the wireframes from the Sneak Peek](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/sneak_peek.md#:~:text=For%20this%20project%20you%20will%20have%20full%20freedom%20in%20terms%20of%20visual%20design%20but%20you%20will%20need%20to%20keep%20the%20following%20wireframes%3A) and our styling.

Thanks for your keen and kind review 😄

### CODE REVIEW

File: db/migrate/20220824174034_add_confirmable_to_devise.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class AddConfirmableToDevise < ActiveRecord::Migration[6.1]
+  def change
+    change_table(:users) do |t| 
+      t.confirmable
```
Comment (GzimAsani) [inline db/migrate/20220824174034_add_confirmable_to_devise.rb]: - [ ] I am getting this error while trying to migrate your database

image

**Review by GzimAsani — CHANGES_REQUESTED** [review]
### Hello @carlylechia           :wave:   
 Good job so far! There are some issues that you still need to work on to go to the next project but you are almost there!
 
 ## Highlights
 * Good job  with your user.rb file :+1: 
 
 ## Required Changes 


- [ ] I was unable to run the migrations properly because of a migration problem shown down on the comment section :arrow_double_down:. The model `user.rb` looks good, but the next reviewer will check if everything works fine on the front end




 ## Optional suggestions
 _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding! clapclapclap
 
 Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
 
 **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
 
 _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20220824174034_add_confirmable_to_devise.rb (hunk 1 above)
Comment (carlylechia) [reply db/migrate/20220824174034_add_confirmable_to_devise.rb]: Hi @GzimAsani,
Everything looks good from my end, so I don't get why the migrations won't run on your computer. Thanks for your review though

Comment (GzimAsani) [reply db/migrate/20220824174034_add_confirmable_to_devise.rb]: Hi @carlylechia, can you please try to clone your project from the beginning and try to run bundle and run migrations, and see if everything works?

File: db/migrate/20220824174034_add_confirmable_to_devise.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class AddConfirmableToDevise < ActiveRecord::Migration[6.1]
+  def change
+    change_table(:users) do |t| 
+      t.confirmable 
+    end
+    add_index  :users, :confirmation_token, :unique => true 
+  end
+end
```
Comment (andersonlebon) [inline db/migrate/20220824174034_add_confirmable_to_devise.rb]: image

- [ ] Kindly address the issue started by the previous reviewer, this is due to some duplicated stuff and errors in migrations. Note that this migration is useless as all the logic of confirmable is done in your `add_devise_to_user` migration. You can solve the error above by removing this migration.

File: db/migrate/20220824185245_add_unconfirmed_to_devise.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      t.string   :unconfirmed_email # Only if using reconfirmable
+  
+        ## Lockable
+      #t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      #t.string   :unlock_token # Only if unlock strategy is :email or :both
+      #t.datetime :locked_at
+      # end
+    end
+  end
+end
```
Comment (andersonlebon) [inline db/migrate/20220824185245_add_unconfirmed_to_devise.rb]: - [ ] Kindly note that all the logic of confirmable was done in the `add_devise_to_user` migration, allow me to say that this migration is useless for now. here is the confirmable already done in the migration `add_devise_to_user` logic :point_down: 

```rb
# frozen_string_literal: true

class AddDeviseToUsers < ActiveRecord::Migration[6.1]
  def self.up
    change_table :users do |t|
      ## Database authenticatable
      t.string :email,              null: false, default: ""
      t.string :encrypted_password, null: false, default: ""

      ## Recoverable
      t.string   :reset_password_token
      t.datetime :reset_password_sent_at

      ## Rememberable
      t.datetime :remember_created_at

      ## Trackable
      # t.integer  :sign_in_count, default: 0, null: false
      # t.datetime :current_sign_in_at
      # t.datetime :last_sign_in_at
      # t.string   :current_sign_in_ip
      # t.string   :last_sign_in_ip

      ## Confirmable
      t.string   :confirmation_token
      t.datetime :confirmed_at
      t.datetime :confirmation_sent_at
      t.string   :unconfirmed_email # Only if using reconfirmable

      ## Lockable
      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
      # t.string   :unlock_token # Only if unlock strategy is :email or :both
      # t.datetime :locked_at


      # Uncomment below if timestamps were not included in your original model.
      # t.timestamps null: false
    end

    add_index :users, :email,                unique: true
    add_index :users, :reset_password_token, unique: true
    # add_index :users, :confirmation_token,   unique: true
    # add_index :users, :unlock_token,         unique: true
  end

  def self.down
    # By default, we don't want to make any assumption about how to roll back a migration when your
    # model already existed. Please edit below which fields you would like to remove in this migration.
    raise ActiveRecord::IrreversibleMigration
  end
end
```

**Review by andersonlebon — CHANGES_REQUESTED** [review]
## Hi @carlylechia 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost-there

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- You  are following GitHub actions :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Your design looks good :100: :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
## Required Changes ♻️

- [ ] In summary, kindly remove the two last migrations to solve the error stated by the previous reviewer and check the reason in the comments under the review.
image


- [ ] Kindly check the comments under the review.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20220824185245_add_unconfirmed_to_devise.rb (hunk 3 above)
Comment (carlylechia) [reply db/migrate/20220824185245_add_unconfirmed_to_devise.rb]: Thank you @andersonlebon, I see what you mean. I have fixed that.
That's two reviews gone out of three 😢 🥺

**Review by andersonlebon — APPROVED** [review]
## Hi @carlylechia 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 :clap: :clap: :100: 

nodding-yes

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 

## Optional suggestions
- [ ] Kindly Check the comments under the review._
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**